rust/sogar/sogar-core/src/sogar_registry.rs

```diff
@@ -0,0 +1,457 @@
+use crate::{parse_digest, Manifest};
+use hyper::body::HttpBody;
+use regex::Regex;
+use saphir::{hyper::body::Buf, prelude::*};
+use serde::Deserialize;
+use slog_scope::{debug, error};
+use std::{collections::HashMap, fs::create_dir_all, io, io::Write, path::Path};
+use tokio_02::io::AsyncWriteExt;
+
+const REPOSITORY: &str = "repository";
+const IMAGE_NAME: &str = "image_name";
+const DIGEST: &str = "digest";
+const TAG: &str = "tag";
+pub const ARTIFACTS_DIR: &str = "artifacts";
+const ARTIFACTS_CONTENT: &str = "content.yaml";
+
+pub const BLOB_PATH: &str = "/registry/v2/<repository>/<image_name>/blobs/<digest>";
+pub const BLOB_GET_LOCATION_PATH: &str = "/registry/v2/<repository>/<image_name>/blobs/uploads";
+pub const UPLOAD_BLOB_PATH: &str = "/registry/<repository>/<image_name>";
+pub const MANIFEST_PATH: &str = "/registry/v2/<repository>/<image_name>/manifests/<tag>";
+
+pub const BLOB_EXIST_ENDPOINT: &str = "is_blob_exist";
+pub const BLOB_GET_LOCATION_ENDPOINT: &str = "get_blob_location";
+pub const BLOB_UPLOAD_ENDPOINT: &str = "save_blob";
+pub const BLOB_DOWNLOAD_ENDPOINT: &str = "get_blob";
+pub const MANIFEST_EXIST_ENDPOINT: &str = "is_manifest_exist";
+pub const MANIFEST_UPLOAD_ENDPOINT: &str = "save_manifest";
+pub const MANIFEST_DOWNLOAD_ENDPOINT: &str = "get_manifest";
+
+const CONTENT_TYPE_HEADER: &str = "content-type";
+const ACCEPT_HEADER: &str = "accept";
+
+pub struct SogarCustomResponse {
+    headers: HashMap<String, String>,
+    status: StatusCode,
+    file: Option<File>,
+}
+
+impl SogarCustomResponse {
+    pub fn new(status: StatusCode) -> Self {
+        SogarCustomResponse {
+            headers: HashMap::new(),
+            status,
+            file: None,
+        }
+    }
+
+    pub fn header(&mut self, key: &str, value: &str) {
+        self.headers.insert(key.to_string(), value.to_string());
+    }
+
+    pub fn file(&mut self, file: File) {
+        self.file = Some(file);
+    }
+}
+
+impl Responder for SogarCustomResponse {
+    fn respond_with_builder(self, builder: Builder, _ctx: &HttpContext) -> Builder {
+        let mut builder_copy = builder;
+        for (key, val) in self.headers {
+            builder_copy = builder_copy.header(&key, val);
+        }
+
+        builder_copy = builder_copy.status(self.status);
+
+        if let Some(file) = self.file {
+            builder_copy = builder_copy.file(file);
+        }
+
+        builder_copy
+    }
+}
+
+pub struct SogarController;
```

Comment:
Nothing prevents us from building a SogarController without calling `SogarController::new` here.
You can add a zero-sized attribute so we need to call `new` and checks are performed.
```suggestion
pub struct SogarController { _priv: () }
```

Reply:
done